### PR TITLE
Add correct error messages to replica.Term and replica.Entries

### PR DESCRIPTION
**Pull request by BramGruneir**
This was motivated by the discussion on #3069 by @bdarnell

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3109)

### CODE REVIEW

File: storage/replica_raftstorage.go [hunk 1]
```diff
@@ -83,6 +83,9 @@ func (r *Replica) InitialState() (raftpb.HardState, raftpb.ConfState, error) {
 // TODO(bdarnell): consider caching for recent entries, if rocksdb's builtin caching
 // is insufficient.
 func (r *Replica) Entries(lo, hi, maxBytes uint64) ([]raftpb.Entry, error) {
+	if lo > hi {
+		return nil, util.Errorf("lo:%d is greater than hi:%d", lo, hi)
```
Comment (mrtracy) [inline storage/replica_raftstorage.go]: Should something like this just be a panic?

Comment (tschottdorf) [inline storage/replica_raftstorage.go]: if you can return an error, why not? the callers will do the panicking.

File: storage/replica_raftstorage.go [hunk 2]
```diff
@@ -112,16 +114,32 @@ func (r *Replica) Entries(lo, hi, maxBytes uint64) ([]raftpb.Entry, error) {
 	// If neither the number of entries nor the size limitations had an
 	// effect, we weren't able to supply everything the client wanted.
 	if len(ents) != int(hi-lo) && (maxBytes == 0 || size < maxBytes) {
```
Comment (mrtracy) [inline storage/replica_raftstorage.go]: I don't quite understand maxBytes; so if a call comes in from `lo` to `hi`, where `lo` (and some number of entries after it) has already been compacted, but there are more than maxBytes worth of entries remaining in between `lo` and `hi`, it's still a valid call?

Comment (bdarnell) [inline storage/replica_raftstorage.go]: The check for `ErrCompacted` should come before the one for `maxBytes`. The original check was kind of sloppy and could cover up missing records if the `maxBytes` limit was hit. The sequence should be
* If `lo` has been compacted, return `ErrCompacted`
* If `ents` contains gaps, return an error (doesn't really matter what)
* If the last entry is not `hi-1` and `size < maxBytes`, return `ErrUnavailable`

Comment (BramGruneir) [inline storage/replica_raftstorage.go]: I'll add some tests around maxbytes to make sure it returns the correct values, even if maxbytes is hit and add the check for gaps/compacted into the scan.  .
And if there are gaps, I'll return ErrUnavailable  .

Comment (BramGruneir) [inline storage/replica_raftstorage.go]: @mrtracy All we really have here, is the interface for the function:
https://github.com/coreos/etcd/blob/master/raft/storage.go

**Comment by BramGruneir** [comment]
I re-wrote a good chunk of entries and added a bunch of new unit tests around maxBytes and adding a gap in the index record.

File: storage/replica_test.go [hunk 3]
```diff
@@ -3365,3 +3366,199 @@ func TestReplicaDestroy(t *testing.T) {
...
+		t.Errorf("17: expected 1 entry, got %d", len(ents))
+	}
+
+	// Case 18: don't hit the gap due to truncation.
+	if _, err := rng.Entries(indexes[4], indexes[9], 0); err != raft.ErrCompacted {
+		t.Errorf("18: expected error %s , got %s", raft.ErrCompacted, err)
+	}
+}
+
+// TestEntires verifies that Entries functions correctly.
```
Comment (bdarnell) [inline storage/replica_test.go]: s/Ent(ri|ir)es)/Term/g

Or just delete the comment; tests only need doc-style comments when you have something non-trivial to say.

**Comment by bdarnell** [comment]
LGTM

File: storage/replica_test.go (hunk 3 above)
Comment (BramGruneir) [inline storage/replica_test.go]: removed the comment.

**Pull request merged**